crates/flipperzero/src/gpio/i2c.rs

```diff
@@ -0,0 +1,402 @@
+
+    /// Reads the 8-bit register at `reg_addr` on `device`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn read_u8(
+        &mut self,
+        device: DeviceAddress,
+        reg_addr: u8,
+        timeout: u32,
+    ) -> Result<u8, Error> {
+        let mut data = 0;
+        if unsafe {
+            sys::furi_hal_i2c_read_reg_8(self.handle, device.0, reg_addr, &mut data, timeout)
+        } {
+            Ok(data)
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Reads the 16-bit register at `reg_addr` on `device`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn read_u16(
+        &mut self,
+        device: DeviceAddress,
+        reg_addr: u8,
+        timeout: u32,
+    ) -> Result<u16, Error> {
+        let mut data = 0;
+        if unsafe {
+            sys::furi_hal_i2c_read_reg_16(self.handle, device.0, reg_addr, &mut data, timeout)
+        } {
+            Ok(data)
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Reads `device`'s memory starting at `mem_addr` into the given buffer.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn read_exact(
+        &mut self,
+        device: DeviceAddress,
+        mem_addr: u8,
+        buf: &mut [u8],
+        timeout: u32,
+    ) -> Result<(), Error> {
+        if unsafe {
+            sys::furi_hal_i2c_read_mem(
+                self.handle,
+                device.0,
+                mem_addr,
+                buf.as_mut_ptr(),
+                buf.len(),
+                timeout,
+            )
+        } {
+            Ok(())
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Writes the given value into the 8-bit register at `reg_addr` on `device`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn write_u8(
+        &mut self,
+        device: DeviceAddress,
+        reg_addr: u8,
+        data: u8,
+        timeout: u32,
+    ) -> Result<(), Error> {
+        if unsafe { sys::furi_hal_i2c_write_reg_8(self.handle, device.0, reg_addr, data, timeout) }
+        {
+            Ok(())
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Writes the given value into the 16-bit register at `reg_addr` on `device`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn write_u16(
+        &mut self,
+        device: DeviceAddress,
+        reg_addr: u8,
+        data: u16,
+        timeout: u32,
+    ) -> Result<(), Error> {
+        if unsafe { sys::furi_hal_i2c_write_reg_16(self.handle, device.0, reg_addr, data, timeout) }
+        {
+            Ok(())
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Writes the given data into `device`'s memory starting at `mem_addr`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn write_all(
+        &mut self,
+        device: DeviceAddress,
+        mem_addr: u8,
+        data: &mut [u8],
+        timeout: u32,
```

Comment:
It would be nice to use `Duration` for the timeouts, rather than bare integers.

Reply:
Slightly annoyingly, the most recent SDK update fixed the documentation for these methods to indicate that these are actually in milliseconds instead of ticks. So using `Duration` would add two conversions (one outside the method, and one inside) on top of the presumed third conversion happening inside the SDK.

Reply:
Inside the SDK it results in a div for an assertion, and a couple of mults. That's also what we end up adding via `Duration`, plus some FFI calls to get the tick frequency twice. Overall it's probably not too costly, so I'll do that and we can improve this via SDK changes later.